marshmallow/fields.py

```diff
@@ -931,62 +931,74 @@ class DateTime(Field):
     }
 
     def __init__(self, format=None, **kwargs):
         super(DateTime, self).__init__(**kwargs)
         # Allow this to be None. It may be set later in the ``_serialize``
         # or ``_desrialize`` methods This allows a Schema to dynamically set the
-        # dateformat, e.g. from a Meta option
-        self.dateformat = format
+        # format, e.g. from a Meta option
+        self.format = format
 
     def _add_to_schema(self, field_name, schema):
         super(DateTime, self)._add_to_schema(field_name, schema)
-        self.dateformat = self.dateformat or schema.opts.dateformat
+        self.format = self.format or \
+            getattr(schema.opts, self.SCHEMA_OPTS_VAR_NAME) or \
+            self.DEFAULT_FORMAT
 
     def _serialize(self, value, attr, obj):
+        self.format = self.format or self.DEFAULT_FORMAT
         if value is None:
             return None
-        dateformat = self.dateformat or self.DEFAULT_FORMAT
-        format_func = self.DATEFORMAT_SERIALIZATION_FUNCS.get(dateformat, None)
+        format_func = self.SERIALIZATION_FUNCS.get(self.format, None)
         if format_func:
             try:
                 return format_func(value, localtime=self.localtime)
-            except (AttributeError, ValueError):
-                self.fail('format', input=value)
+            except (TypeError, AttributeError, ValueError):
+                self.fail('format', input=value, obj_type=self.OBJ_TYPE)
         else:
-            return value.strftime(dateformat)
+            return value.strftime(self.format)
 
     def _deserialize(self, value, attr, data):
+        self.format = self.format or self.DEFAULT_FORMAT
         if not value:  # Falsy values, e.g. '', None, [] are not valid
-            raise self.fail('invalid')
-        dateformat = self.dateformat or self.DEFAULT_FORMAT
-        func = self.DATEFORMAT_DESERIALIZATION_FUNCS.get(dateformat)
+            raise self.fail('invalid', obj_type=self.OBJ_TYPE)
+        func = self.DESERIALIZATION_FUNCS.get(
+            self.format,
+        )
         if func:
             try:
                 return func(value)
             except (TypeError, AttributeError, ValueError):
-                raise self.fail('invalid')
-        elif dateformat:
+                raise self.fail('invalid', obj_type=self.OBJ_TYPE)
+        elif self.format:
             try:
-                return dt.datetime.strptime(value, dateformat)
+                return dt.datetime.strptime(value, self.format)
             except (TypeError, AttributeError, ValueError):
-                raise self.fail('invalid')
+                raise self.fail('invalid', obj_type=self.OBJ_TYPE)
         elif utils.dateutil_available:
             try:
-                return utils.from_datestring(value)
-            except TypeError:
-                raise self.fail('invalid')
+                parsed = utils.from_datestring(value)
+                return dt.datetime(
```

Comment:
I don't understand this change. I don't think this would return a `dt.date` when called from `Date`.

Isn't this code unreachable anyway ? (https://github.com/marshmallow-code/marshmallow/issues/758)

(I don't mean to throw #758 in your way, this is another story. If the code is indeed unreachable, which I think it is, don't worry too much about it. I just don't understand the change.)

Reply:
The object creation is an oversight. I'll fix that. I think #758 is beyond the scope of this PR. I can look at it after this change, though, if you need someone to work on it.